### PR TITLE
[Cancun][Ethereum.Tests] Add Pyspec as an independent project with fixtures being uploaded manually there.

### LINKED ISSUE #6166: EthereumTests.sln (silently) fails to execute any test fixtures from execution-spec-tests
**Description**
Executing:
```
dotnet test Ethereum.Blockchain.Block.Test/Ethereum.Blockchain.Block.Test.csproj
```
fails to find any of the [ethereum/tests](https://github.com/ethereum/tests) test fixtures that originate from the [ethereum/execution-spec-tests](https://github.com/ethereum/execution-spec-tests) repository. This is likely due to directory renaming (that took place in June) within execution-spec-tests, but probably existed in the propagated fixtures in ethereum/tests for much longer.

**Steps to Reproduce**
Use master (6472650d2c) or cancun (45c4cfcc02) and filter to reduce noise:
```
dotnet test Ethereum.Blockchain.Block.Test/Ethereum.Blockchain.Block.Test.csproj --filter FullyQualifiedName~Ethereum.Blockchain.Block.Test.Pyspecs.VMTests
```

**Actual behavior**
Currently, no tests are ran and the command reports:
```
No test matches the given testcase filter `FullyQualifiedName~Ethereum.Blockchain.Block.Test.Pyspecs.VMTests` in /home/dtopz/code/github/NevermindETH/nethermind/src/Nethermind/Ethereum.Blockchain.Block.Test/bin/Debug/net7.0/Ethereum.Blockchain.Block.Test.dll
```

Modifying the path to a directory that contains JSON fixtures resolves the issue, but probably it'd be desirable to search directories recursively under the provided directory to find all JSON fixtures in the sub-directories under the provided directory. Modifying one of the test files as following works, for example:
```diff
diff --git a/src/Nethermind/Ethereum.Blockchain.Block.Test/Pyspecs/VMTests.cs b/src/Nethermind/Ethereum.Blockchain.Block.Test/Pyspecs/VMTests.cs
index 9bdcec07bd..45d4419f5b 100644
--- a/src/Nethermind/Ethereum.Blockchain.Block.Test/Pyspecs/VMTests.cs
+++ b/src/Nethermind/Ethereum.Blockchain.Block.Test/Pyspecs/VMTests.cs
@@ -17,7 +17,7 @@ public class VMTests : BlockchainTestBase
 
     public static IEnumerable<BlockchainTest> LoadTests()
     {
-        var loader = new TestsSourceLoader(new LoadBlockchainTestsStrategy(), "Pyspecs/vm");
+        var loader = new TestsSourceLoader(new LoadBlockchainTestsStrategy(), "Pyspecs/frontier/opcodes");
 
         return (IEnumerable<BlockchainTest>)loader.LoadTests();
     }
```

**Expected behavior**
Tests should run and generate the following output:
```
Running 009-fork=Shanghai at 14:14:21.067649
Running 007-fork=London at 14:14:21.067634
Running 006-fork=Berlin at 14:14:21.067589
Running 005-fork=Istanbul at 14:14:21.067584
Running 002-fork=Byzantium at 14:14:21.067614
Running 008-fork=Merge at 14:14:21.067622
Running 004-fork=ConstantinopleFix at 14:14:21.067614
Running 003-fork=Constantinople at 14:14:21.067625
Running 000-fork=Frontier at 14:14:21.067639
Running 001-fork=Homestead at 14:14:21.067617

Passed!  - Failed:     0, Passed:    10, Skipped:     0, Total:    10, Duration: 580 ms - Ethereum.Blockchain.Block.Test.dll (net7.0)
```

**Desktop (please complete the following information):**
Please provide the following information regarding your setup:
 - Operating System: Linux
 - Version: Ubuntu 22.04
 - Installation Method: Built on command-line.
 - Consensus Client: None

**Additional context**
Background: I help maintain execution-spec-tests. I'd love to see Nethermind use the execution-spec-test fixtures directly from the repo's releases instead of via ethereum/tests. Currently these are available on the repo's [Github releases page](https://github.com/ethereum/execution-spec-tests/releases/).
(You'd need to use `fixtures_develop.gar.gz`: https://github.com/ethereum/execution-spec-tests/releases/download/v1.0.5/fixtures_develop.tar.gz). We're happy to discuss the best way for you get these artifacts - please reach out. But currently using ethereum/tests will result in a slower feedback loop and a delay in between the execution-spec-tests release and when they'd be available for use via ethereum/tests.

### PR DESCRIPTION
**Pull request by smartprogrammer93**
Resolves #6166

_Please choose one of the keywords above to refer to the issue this PR solves followed by the issue number (e.g. Fixes #000). If no issue number, remove the line. Also, remove everything marked optional that is not applicable. Remove this note after reading._

## Changes

- Add pytest project
- add extra message to indicate specname resolved while running the test
- fix incorrect cancun address

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [x] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by rubo — COMMENTED** [review]
I have mixed feelings about the inclusion of so many JSON files. I'd instead prefer downloading them in the workflow.

**Comment by smartprogrammer93** [comment]
@rubo 
We can have the fixtures be downloaded from the release. but we need to update this url in code manually to not have problems with tests every time ef devops update their tests. we are not always passing all of these. it would fail a lot of PRs. 
I can have the `TestsSourceLoader` pull them from the url and unpack them. but the bigger they get the uglier running the tests become. I am not a big fan of this strat. I prefer uploading them ourselves in our repo

**Comment by LukaszRozmej** [comment]
Where those test come from, arn't they already in some repo we can submodule like with other Ethereum tests?

**Comment by smartprogrammer93** [comment]
@LukaszRozmej 
From here:
https://github.com/ethereum/execution-spec-tests/releases
fixtures_develop one

**Comment by LukaszRozmej** [comment]
> @LukaszRozmej From here: https://github.com/ethereum/execution-spec-tests/releases fixtures_develop one you can check the original issue here #6166

So those tests should be brought in as a submodule, not coping them here

**Comment by smartprogrammer93** [comment]
> So those tests should be brought in as a submodule, not coping them here
@LukaszRozmej how would you bring those tests as a submodule if they are only available in the release not in the repo itself. they do end up in ethereum/tests at some point. but after someone from EF DevOps moves them there. EF DevOps requested that we run pyspec tests directly from release here:
https://discord.com/channels/595666850260713488/910910348922589184/1159862930637078650

**Comment by danceratopz** [comment]
> > So those tests should be brought in as a submodule, not coping them here
> 
> @LukaszRozmej how would you bring those tests as a submodule if they are only available in the release not in the repo itself. they do end up in ethereum/tests at some point. but after someone from EF DevOps moves them there. EF DevOps requested that we run pyspec tests directly from release here: https://discord.com/channels/595666850260713488/910910348922589184/1159862930637078650

Hi @LukaszRozmej, the test fixtures in question are generated programmatically by the test cases and framework provided by ethereum/execution-spec-tests. As they're generated artifacts, we'd prefer to not check the fixtures in the repo, but are open to find a good solution for client teams. This repo is a relatively new approach to generating test fixtures and mainly targets tests for features under active development.

The maintainer of ethereum/tests includes these fixtures in ethereum/tests releases, but these aren't released at the same frequency as execution-spec-tests, so I suggested retrieving the execution-spec-tests fixtures directly for faster feedback (in #6166, "Additional Context").

> We can have the fixtures be downloaded from the release. but we need to update this url in code manually to not have problems with tests every time ef devops update their tests

@smartprogrammer93 Regarding manual update, Github doesn't provide a static link to the latest release artifacts, as far as I can tell, but something along these line works:

```bash
#!/bin/bash

# requires jq
# sudo apt install jq

# The following two artifacts are intended for consumption by clients:
# - fixtures.tar.gz: Generated up to the last deployed fork.
# - fixtures_develop.tar.gz: Generated up to and including the latest dev fork.
# As of Oct 2023, dev is Cancun, deployed is Shanghai.

ARTIFACT="fixtures_develop.tar.gz"  

OWNER="ethereum"
REPO="execution-spec-tests"

DOWNLOAD_URL=$(curl -s https://api.github.com/repos/$OWNER/$REPO/releases/latest \
                   | jq -r '.assets[] | select(.name=="'$ARTIFACT'").browser_download_url')

# Sanity check for the download url: contains a version tag prefixed with "v"
if [[ "$DOWNLOAD_URL" =~ v[0-9]+\.[0-9]+\.[0-9]+ ]]; then
    curl -LO $DOWNLOAD_URL
else
    echo "Error: URL does not contain a valid version tag (URL: ${DOWNLOAD_URL})."
    exit 1
fi
```

**Comment by LukaszRozmej** [comment]
> Hi @LukaszRozmej, the test fixtures in question are generated programmatically by the test cases and framework provided by ethereum/execution-spec-tests. As they're generated artifacts, we'd prefer to not check the fixtures in the repo, but are open to find a good solution for client teams. This repo is a relatively new approach to generating test fixtures and mainly targets tests for features under active development.

How long does it take to generate them? If it is fast maybe better would be to:
a) Check if they were generated relatively recently under some temp/internal path
b) If not regenerate them
c) Run from temp/internal path

Rather than keeping them locally.

To be honest a) & b) can be also changed to just downloading them from github release.

This can be done even under unit test.

**Comment by danceratopz** [comment]
> How long does it take to generate them? 

~30 seconds (including Cancun). This excludes time to build/retrieve external deps. This time will increase with every new fork as test cases are added.

> If it is fast maybe better would be to: a) Check if they were generated relatively recently under some temp/internal path b) If not regenerate them c) Run from temp/internal path

I like the idea, but the external dependencies prevent this from being a reasonable approach, imo. Currently, generating fixtures for development features require building a development branch of geth (for `evm t8n`). 

But in the future, fixtures for a release will be generated from multiple branches of different clients' `evm t8n` tools. For example, evmone for EOF or [ethereum/execution-specs](https://github.com/ethereum/execution-specs) for EIPs implemented via EELS. 

I think it's better that we handle this complexity in one place, so you don't have to worry about it.

> Or a') Check the version of github release with current downloaded version b') If github newer or local missing - download and extract to temp/internal path c) Run from temp/internal path

I think this is a good approach. If it would help to publish the artifacts additionally elsewhere, we would certainly consider that!

**Comment by flcl42** [comment]
What if we make a repo for just for the generation results with an action that is responsible for fixture generation and making commit with the fixtures. We add it as a submodule to the client repo and update on demand or with an action too

**Comment by smartprogrammer93** [comment]
> What if we implemented some change that is new and does not follow the tests by purpose?

We can't really implement a change that breaks those tests unless the tests are old, and then we disable them until renewed. Once they are renewed, we update the release url to be downloaded and enable them.

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        if (!string.IsNullOrEmpty(versionAndDir[0]))
+            archiveVersion = versionAndDir[0];
+        if (!string.IsNullOrEmpty(versionAndDir[1]))
+            archiveName = versionAndDir[1];
+        if (!string.IsNullOrEmpty(versionAndDir[2]))
+            testsDir = versionAndDir[2];
+        string testsDirectoryName = $"{AppContext.BaseDirectory}PyTests/{archiveVersion}/{archiveName.Split('.')[0]}";
+        if (!Directory.Exists(testsDirectoryName))
+        {
+            HttpClient httpClient = new();
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:35]: using

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            HttpResponseMessage response = httpClient.GetAsync(string.Format(Constants.ARCHIVE_URL_TEMPLATE, archiveVersion, archiveName)).GetAwaiter().GetResult();
+            response.EnsureSuccessStatusCode();
+            using Stream contentStream = response.Content.ReadAsStreamAsync().GetAwaiter().GetResult();
+            using TarArchive archive = TarArchive.Open(contentStream);
+            archive.ExtractToDirectory(testsDirectoryName);
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:40]: Maybe lets move it to separate method

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        if (versionAndDir.Length < 3)
+        {
+            throw new ArgumentException("Invalid testsDir argument. It should be in format: <version>,<archiveName>,<testsDir>");
+        }
+        if (!string.IsNullOrEmpty(versionAndDir[0]))
+            archiveVersion = versionAndDir[0];
+        if (!string.IsNullOrEmpty(versionAndDir[1]))
+            archiveName = versionAndDir[1];
+        if (!string.IsNullOrEmpty(versionAndDir[2]))
+            testsDir = versionAndDir[2];
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:31]: Can we extract this to a method?

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/ShanghaiTests.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+[TestFixture]
+[Parallelizable(ParallelScope.All)]
+public class ShanghaiTests : BlockchainTestBase
+{
+    [TestCaseSource(nameof(LoadTests))]
+    public async Task Test(BlockchainTest test) => await RunTest(test);
+
+    private static IEnumerable<BlockchainTest> LoadTests()
+    {
+        TestsSourceLoader loader = new(new LoadPyspecTestsStrategy(), ",,fixtures/shanghai");
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/ShanghaiTests.cs:20]: I know you want to pass 3 params, though `",,fixtures/shanghai"` looks weird.

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            HttpClient httpClient = new();
+            HttpResponseMessage response = httpClient.GetAsync(string.Format(Constants.ARCHIVE_URL_TEMPLATE, archiveVersion, archiveName)).GetAwaiter().GetResult();
+            response.EnsureSuccessStatusCode();
+            using Stream contentStream = response.Content.ReadAsStreamAsync().GetAwaiter().GetResult();
+            using TarArchive archive = TarArchive.Open(contentStream);
+            archive.ExtractToDirectory(testsDirectoryName);
+        }
+        IEnumerable<string> testDirs = !string.IsNullOrEmpty(testsDir)
+            ? Directory.EnumerateDirectories(testsDirectoryName + "/" + testsDir, "*", new EnumerationOptions { RecurseSubdirectories = true })
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:43]: use Path.Combine for better OS compatibility

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        {
+            throw new ArgumentException("Invalid testsDir argument. It should be in format: <version>,<archiveName>,<testsDir>");
+        }
+        if (!string.IsNullOrEmpty(versionAndDir[0]))
+            archiveVersion = versionAndDir[0];
+        if (!string.IsNullOrEmpty(versionAndDir[1]))
+            archiveName = versionAndDir[1];
+        if (!string.IsNullOrEmpty(versionAndDir[2]))
+            testsDir = versionAndDir[2];
+        string testsDirectoryName = $"{AppContext.BaseDirectory}PyTests/{archiveVersion}/{archiveName.Split('.')[0]}";
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:32]: use Path.Combine

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        IEnumerable<string> testDirs = !string.IsNullOrEmpty(testsDir)
+            ? Directory.EnumerateDirectories(testsDirectoryName + "/" + testsDir, "*", new EnumerationOptions { RecurseSubdirectories = true })
+            : Directory.EnumerateDirectories(testsDirectoryName, "*", new EnumerationOptions { RecurseSubdirectories = true });
+        List<BlockchainTest> testJsons = new();
+        foreach (string testDir in testDirs)
+        {
+            testJsons.AddRange(LoadTestsFromDirectory(testDir, wildcard));
+        }
+        return testJsons;
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:50]: Linq SelectMany?

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    private IEnumerable<BlockchainTest> LoadTestsFromDirectory(string testDir, string wildcard)
+    {
+        List<BlockchainTest> testsByName = new();
+        IEnumerable<string> testFiles = Directory.EnumerateFiles(testDir);
+
+        foreach (string testFile in testFiles)
+        {
+            FileTestsSource fileTestsSource = new(testFile, wildcard);
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:60]: Does it make sense to pass this custom wildcard here?

**Review by LukaszRozmej — COMMENTED** [review]
Do you know if the directory will be preserved/cached on GithubActions machines or downloaded every time?

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs (hunk 8 above)
Comment (smartprogrammer93) [reply src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:65]: Yeah i guess so, maybe you want to run a specific test. No?

**Comment by smartprogrammer93** [comment]
@LukaszRozmej it will download only once per github action execution. As long as the directory with the fixtures is not deleted, it will not download again. (given the test uses the same version and archive name)

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    public IEnumerable<IEthereumTest> Load(string testsDir, string wildcard = null)
+    {
+        testsDir = GetArchiveVersionAndName(testsDir, out string archiveVersion, out string archiveName);
+        string testsDirectoryName = Path.Combine(AppContext.BaseDirectory, "PyTests", archiveVersion, archiveName.Split('.')[0]);
+        if (!Directory.Exists(testsDirectoryName)) // Prevent redownloading the fixtures if they already exists with this version and archive name
+            DownloadAndExtract(archiveVersion, archiveName, testsDirectoryName);
+        IEnumerable<string> testDirs = !string.IsNullOrEmpty(testsDir)
+            ? Directory.EnumerateDirectories(Path.Combine(testsDirectoryName, testsDir), "*", new EnumerationOptions { RecurseSubdirectories = true })
+            : Directory.EnumerateDirectories(testsDirectoryName, "*", new EnumerationOptions { RecurseSubdirectories = true });
+        return testDirs.SelectMany(td => LoadTestsFromDirectory(td, wildcard)).ToList();
```
Comment (LukaszRozmej) [inline src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:27]: Actually you can probably drop ToList and lazy load them, might be even better.

File: src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs (hunk 8 above)
Comment (LukaszRozmej) [reply src/Nethermind/Ethereum.Blockchain.Pyspec.Test/LoadPyspecTestsStrategy.cs:65]: But the wildcard would be here `",,fixtures/shanghai"`, does it makes any sense for `FileTestsSource`? I think not?

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**